qiskit/transpiler/passes/scheduling/alap.py

```diff
@@ -117,26 +110,23 @@ def run(self, dag):
                     t1 = t0 + op_duration
 
             for bit in node.qargs:
-                delta = t0 - idle_before[bit]
-                if delta > 0:
-                    new_dag.apply_operation_front(Delay(delta, time_unit), [bit], [])
                 idle_before[bit] = t1
 
-            new_dag.apply_operation_front(node.op, node.qargs, node.cargs)
+            # Place the node in time slot. Ignore delay since this is identical to empty.
+            if not isinstance(node.op, Delay):
+                # Note that some transformation pass can override .op attribute of DAGOpNode
+                # without breaking the reference to object. However, once operation is updated,
+                # there is no guarantee that the duration of node is unchanged.
+                key = node.name, node.sort_key, getattr(node, "_node_id")
+                time_slot[key] = t1
 
+        # Compute maximum instruction available time, i.e. very end of t1
         circuit_duration = max(idle_before.values())
-        for bit, before in idle_before.items():
-            delta = circuit_duration - before
-            if not (delta > 0 and isinstance(bit, Qubit)):
-                continue
-            new_dag.apply_operation_front(Delay(delta, time_unit), [bit], [])
-
-        new_dag.name = dag.name
-        new_dag.metadata = dag.metadata
-        new_dag.calibrations = dag.calibrations
-
-        # set circuit duration and unit to indicate it is scheduled
-        new_dag.duration = circuit_duration
-        new_dag.unit = time_unit
 
-        return new_dag
+        # Sort time slot by t0. Note that ALAP pass is inversely schedule, thus
+        # t0 is computed by subtracting entire circuit duration from t1.
+        self.property_set["time_slot"] = {
+            n: circuit_duration - t1
+            for n, t1 in sorted(time_slot.items(), key=lambda kv: kv[1], reverse=True)
```

Comment:
The order doesn't matter here because `times_slot` is dictionary keyed on the node. I can remove this sort if eventually the overhead becomes an issue.

Reply:
I remove this redundant sort in https://github.com/Qiskit/qiskit-terra/pull/7709/commits/5ae16299e935014f1b94106127c5176c39fd6e33
The motivation of sort is to make the dict human readable but anyways this data is not exposed to the users.